### DISCUSSION TITLE
Routing improvements

### DISCUSSION THREAD

**Discussion by vytas7** [falconry/falcon#1818] [Ideas]
Improved routing is undeniably a popular topic among feature requests from the community.
Let us try to start a discussion here on _Discussions_ :wave:

### Tracked issues
* Explore object-based routing https://github.com/falconry/falcon/issues/319
* Advanced filters such as matching multiple path segments etc https://github.com/falconry/falcon/issues/423
* Routing all paths to one handler function (**editor's note**: potential duplication of the sink feature) https://github.com/falconry/falcon/issues/648
* Conflicting URI templates https://github.com/falconry/falcon/issues/663
* Add support for query string parameters in route templates https://github.com/falconry/falcon/issues/838
* Cascade blueprints a-la Flask https://github.com/falconry/falcon/issues/855
* Calling API from inside Resource handler (loosely related) https://github.com/falconry/falcon/issues/1228
* Inter-resource delegation, similar to Flask's `url_for()` https://github.com/falconry/falcon/issues/1264
* Allow different parameter names for the same segment across routes https://github.com/falconry/falcon/issues/1471
* Use pattern hints from converters https://github.com/falconry/falcon/issues/1567
* Include responder in middleware process_resource https://github.com/falconry/falcon/issues/1629
* (Probably a couple more)

### Workarounds
* Using [sinks](https://falcon.readthedocs.io/en/stable/api/api.html#falcon.API.add_sink) to match any arbitrary path regex.
* [Rewriting request path in middleware](https://falcon.readthedocs.io/en/stable/api/middleware.html#middleware) before routing (think internal redirect) in order to match other routes or sinks.
* Implementing a [custom router](https://falcon.readthedocs.io/en/stable/api/routing.html#custom-routers).
* Anything else?

We would love to hear which potential routing improvements are the most important to You, the framework user.
Feel free to share ideas for improvement, examples from your use cases, and workarounds you found!

**Comment by vytas7** [comment]
Reposting from Gitter
> **Morteza NourelahiAlamdari** @mortymacs 10:50
> I was trying to have repetitive routes in Falcon and I didn't find anything related to it. For example something like `<parent>/folders/<child>/folders/<parent>/folders/<child>/`....
I have an idea in my head and I want to know if we don't have this feature in Falcon, I can start implementing it. It's useful when we want to address nested patterns. Is there any better approach that exists for it?

**Reply by davetapley** [reply]
@vytas7 I'm running in to exactly this right now, I don't suppose there's a good solution yet?

I only care about nesting one level or n, I hoped I could use `suffix=` and fall through, but I get `UnacceptableRouteError` 😞

**Reply by vytas7** [reply]
@davetapley I think this was about unlimited recursion, which may be conceptually more complex. Just unrolling into +1 to _n_ different routes may be easier, although as I understand it didn't work as expected.
Could you give an example of how these routes looked like when you tried to use a suffix?

In either case, you could use a path converter to match a tail including `/`, and then regex the rest yourself in middleware or a lower-level responder, but I understand it gets ugly quick if you have more such routes.

**Comment by vytas7** [comment]
More from Gitter
> **Morteza NourelahiAlamdari** @mortymacs
> Is it possible to get a route "suffix" in middleware? like in process_resource?

This is apparently already tracked as https://github.com/falconry/falcon/issues/1629 (added to the introductory list too).

**Comment by adsahay** [comment]
I'd like to vote for this - either object based or flask-blueprint like routing. It helps for larger projects. This could also be an add-on as you would want to keep everything backward compatible. If I get some direction I am happy to attempt a proof of concept.

**Comment by B1gG** [comment]
Hi guys,
I am playing with add_static_route to serve some static files while developing as in production it should be captured by nginx.
But I was wondering should be there a way to skip the middleware processing for static routes?
at the moment I keep repeating this thing in each middleware methods:
```
        if '/static/images' in req.path:
            return
```

**Comment by kgriffs** [comment]
Is there an existing issue we can link for implementing something along the lines of `url_for` for Falcon?

**Reply by vytas7** [reply]
Yes. there is: https://github.com/falconry/falcon/issues/1264 (added to the list).

**Reply by kgriffs** [reply]
Thanks for tracking that down. I updated the title/description for the issue.

**Comment by aprilahijriyan** [comment]
Hi @vytas7 , I just created a routing sub application on my project maybe it could be part of falcon?

To add a sub-application:

https://github.com/aprilahijriyan/falca/blob/develop/falca/scaffold.py#L105-L121

Main logic:

WSGI:
https://github.com/aprilahijriyan/falca/blob/develop/falca/app.py#L17-L21

ASGI:
https://github.com/aprilahijriyan/falca/blob/develop/falca/app.py#L34-L38

**Comment by maxking** [comment]
We've been using a Object router in Mailman 3 that we have a [custom implementation](https://gitlab.com/mailman/mailman/-/blob/master/src/mailman/rest/wsgiapp.py#L85).

It could use more documentation, but in terms of usage, we define a Top level resource `Root` [here](https://gitlab.com/mailman/mailman/-/blob/master/src/mailman/rest/root.py#L47) and then we use a `@child` decorator to define a sub-resource. Each method gets passed a list of URL segments to make decision on routing further in the depth. 

I thought I'd share it here as I came across this discussion form Gitter.

This is what is mentioned in https://github.com/falconry/falcon/issues/319

**Reply by vytas7** [reply]
Thanks for sharing @maxking!
